In [1]:
print("Hello, MySQL Notebook!")

Hello, MySQL Notebook!


# SQL-Based Data Analysis using MySQL

## Celebal Technologies Internship - Assignment 2

**Name:** Prachurya Nanda

**Dataset:** Sample Superstore

**Database:** MySQL

### Objective
Analyze sales data using SQL filtering, aggregation, sorting, business queries and validation techniques on the Sample Superstore dataset.

In [2]:
import pandas as pd
import mysql.connector

In [3]:
connection = mysql.connector.connect(
    host="localhost",
    user="root",
    password="gayPrince8$",
    database="superstore_db"
)

print("Connected Successfully!")

Connected Successfully!


In [4]:
query = """
SELECT *
FROM superstore
LIMIT 10;
"""

df = pd.read_sql(query, connection)

df

C:\Users\prapt\AppData\Local\Temp\ipykernel_268540\3655285333.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, connection)


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11-08-2016,11-11-2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11-08-2016,11-11-2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,06-12-2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10-11-2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10-11-2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164
5,6,CA-2014-115812,06-09-2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-FU-10001487,Furniture,Furnishings,Eldon Expressions Wood and Plastic Desk Access...,48.8600,7,0.00,14.1694
6,7,CA-2014-115812,06-09-2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AR-10002833,Office Supplies,Art,Newell 322,7.2800,4,0.00,1.9656
7,8,CA-2014-115812,06-09-2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.1520,6,0.20,90.7152
8,9,CA-2014-115812,06-09-2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-BI-10003910,Office Supplies,Binders,DXL Angle-View Binders with Locking Rings by S...,18.5040,3,0.20,5.7825
9,10,CA-2014-115812,06-09-2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,OFF-AP-10002892,Office Supplies,Appliances,Belkin F5C206VTEL 6 Outlet Surge,114.9000,5,0.00,34.4700


### Observation

The dataset contains customer, product, sales, quantity, discount and profit information for retail sales transactions. The first 10 records were displayed successfully.

In [5]:
query = """
SELECT COUNT(*) AS Total_Records
FROM superstore;
"""

pd.read_sql(query, connection)

C:\Users\prapt\AppData\Local\Temp\ipykernel_268540\580848000.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql(query, connection)


,Total_Records
0,9694


In [6]:
def run_query(query):
    return pd.read_sql(query, connection)

In [7]:
run_query("""
SELECT
    ROUND(SUM(Sales),2) AS Total_Sales,
    ROUND(SUM(Profit),2) AS Total_Profit,
    SUM(Quantity) AS Total_Quantity
FROM superstore;
""")

C:\Users\prapt\AppData\Local\Temp\ipykernel_268540\1537842415.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  return pd.read_sql(query, connection)


,Total_Sales,Total_Profit,Total_Quantity
0,2272449.86,282857.75,36749.0


### Observation

The dataset contains 9,694 records. The above query summarizes the overall sales, profit, and quantity sold, providing a quick overview of business performance.

In [8]:
run_query("""
SELECT
    COUNT(DISTINCT Category) AS Categories,
    COUNT(DISTINCT Region) AS Regions,
    COUNT(DISTINCT Segment) AS Segments
FROM superstore;
""")

C:\Users\prapt\AppData\Local\Temp\ipykernel_268540\1537842415.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  return pd.read_sql(query, connection)


,Categories,Regions,Segments
0,3,4,3


### Observation

The dataset contains 3 product categories, 4 sales regions, and 3 customer segments.

In [9]:
run_query("""
SELECT
    Region,
    ROUND(SUM(Sales),2) AS Total_Sales
FROM superstore
GROUP BY Region
ORDER BY Total_Sales DESC;
""")

C:\Users\prapt\AppData\Local\Temp\ipykernel_268540\1537842415.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  return pd.read_sql(query, connection)


,Region,Total_Sales
0,West,713471.34
1,East,672194.05
2,Central,497800.87
3,South,388983.59


### Observation

The West region recorded the highest total sales among all regions.

In [10]:
print("Sales by Category")
display(run_query("""
SELECT
    Category,
    ROUND(SUM(Sales),2) AS Total_Sales
FROM superstore
GROUP BY Category
ORDER BY Total_Sales DESC;
"""))

print("\nTop 10 Products by Sales")
display(run_query("""
SELECT
    `Product Name`,
    ROUND(SUM(Sales),2) AS Total_Sales
FROM superstore
GROUP BY `Product Name`
ORDER BY Total_Sales DESC
LIMIT 10;
"""))

print("\nTop 5 Customers by Sales")
display(run_query("""
SELECT
    `Customer Name`,
    ROUND(SUM(Sales),2) AS Total_Sales
FROM superstore
GROUP BY `Customer Name`
ORDER BY Total_Sales DESC
LIMIT 5;
"""))

Sales by Category


C:\Users\prapt\AppData\Local\Temp\ipykernel_268540\1537842415.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  return pd.read_sql(query, connection)


,Category,Total_Sales
0,Technology,835900.07
1,Furniture,733046.86
2,Office Supplies,703502.93



Top 10 Products by Sales


C:\Users\prapt\AppData\Local\Temp\ipykernel_268540\1537842415.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  return pd.read_sql(query, connection)


,Product Name,Total_Sales
0,Canon imageCLASS 2200 Advanced Copier,61599.82
1,Fellowes PB500 Electric Punch Plastic Comb Bin...,27453.38
2,Cisco TelePresence System EX90 Videoconferenci...,22638.48
3,HON 5400 Series Task Chairs for Big and Tall,21870.58
4,GBC DocuBind TL300 Electric Binding System,19823.48
5,GBC Ibimaster 500 Manual ProClick Binding System,19024.50
6,Hewlett Packard LaserJet 3310 Copier,18839.69
7,HP Designjet T520 Inkjet Large Format Printer ...,18374.90
8,GBC DocuBind P400 Electric Binding System,17965.07
9,High Speed Automatic Electric Letter Opener,17030.31


C:\Users\prapt\AppData\Local\Temp\ipykernel_268540\1537842415.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  return pd.read_sql(query, connection)



Top 5 Customers by Sales


,Customer Name,Total_Sales
0,Sean Miller,25043.05
1,Tamara Chand,19017.85
2,Raymond Buch,15117.34
3,Tom Ashbrook,14595.62
4,Adrian Barton,14355.61


### Observation

Technology generated the highest total sales among all product categories. The Canon imageCLASS 2200 Advanced Copier was the top-selling product, while Sean Miller recorded the highest total sales among customers.

In [11]:
print("Data Validation")
display(run_query("""
SELECT
    COUNT(*) AS Total_Rows,
    SUM(CASE WHEN Sales IS NULL THEN 1 ELSE 0 END) AS Null_Sales,
    SUM(CASE WHEN Profit IS NULL THEN 1 ELSE 0 END) AS Null_Profit
FROM superstore;
"""))

print("\nDuplicate Order IDs")
display(run_query("""
SELECT
    `Order ID`,
    COUNT(*) AS Frequency
FROM superstore
GROUP BY `Order ID`
HAVING COUNT(*) > 1
LIMIT 10;
"""))

Data Validation


C:\Users\prapt\AppData\Local\Temp\ipykernel_268540\1537842415.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  return pd.read_sql(query, connection)


,Total_Rows,Null_Sales,Null_Profit
0,9694,0.0,0.0



Duplicate Order IDs


C:\Users\prapt\AppData\Local\Temp\ipykernel_268540\1537842415.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  return pd.read_sql(query, connection)


,Order ID,Frequency
0,CA-2016-152156,2
1,US-2015-108966,2
2,CA-2014-115812,7
3,US-2015-118983,2
4,CA-2014-143336,3
5,CA-2016-137330,2
6,CA-2016-121755,2
7,US-2015-150630,6
8,CA-2016-117590,2
9,CA-2015-117415,4


### Observation

The dataset contains 9,694 records with no NULL values in the Sales or Profit columns. Duplicate Order IDs represent multiple products within the same customer order and are expected in transactional sales data.